### PR TITLE
feat(fw): run evm blocktest on generated fixtures

### PR DESCRIPTION
**Pull request by danceratopz**
## 🗒️ Description
This PR optionally enables the execution of `evm blocktest` on each generated JSON fixture file. The check is ran as a function in the `FixtureCollector` in a module-scope pytest fixture (so is not executed per test function or per test-case).

If `--t8n-dump-dir` is additionally specified, the output of the `evm blocktest` command is also written there along with a `verify_fixtures.sh` script to re-run the command.

There's a couple of other changes to the t8n-dump directories:
- The JSON fixtures always get copied to the corresponding t8n-dump directory.
- The t8n-dump directory structure:
    -  Adds an additional directory level for the output of the evm test command and a copy of the json fixtures. 
    - Stays as flat as possible, but the base directory is now named to match the location of the test case within ./tests.
    ```
        /tmp/evmone/berlin__eip2930_access_list__test_acl__test_access_list/
        ├── fork_Berlin
        │   ├── 0
        │   │   ├── args.py
        │   │   ├── input
        │   │   │   ├── alloc.json
        │    ....
        │   ├── test_access_list.json
        │   ├── verify_fixtures_args.py
        │   ├── verify_fixtures_returncode.txt
        │   ├── verify_fixtures.sh
        │   ├── verify_fixtures_stderr.txt
        │   └── verify_fixtures_stdout.txt
         ...
        ```
### Examples:
1. No fixture checking performed
    ```
    fill
    ```
2. Use first "evm" binary in path as a t8n tool and as an evm *test tool (same binary used)
    ```
    fill --verify-fixtures
    ```
3. Use a different evm binary as an evm test tool (first evm binary in path used as t8n tool; --verify-fixtures not necessary)
    ```
    fill --verify-fixtures-bin=~/code/github/marioevz/go-ethereum/build/bin/evm
    ```
4. Explicitly set both t8n and test tools; dump debug data
    ```
    fill --evm-bin=~/code/github/ethereum/evmone/build/bin/evmone-t8n --verify-fixtures-bin=~/code/github/marioevz/go-ethereum/build/bin/evm  --evm-dump-dir=/tmp/evmone
    ```

## 🔗 Related Issues
- Partially implements #237 (just need to add the calls with relevant command-line arguments in tox or ci/cd).
- Makes #185 partially redundant.

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [x] All: Considered squashing commits to improve commit history.
- [x] All: Added an entry to [CHANGELOG.md](../docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](../docs/) directory.
- [x] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

**Review by marioevz — COMMENTED** [review]
Looks very good! Just a few comments to keep in mind.

File: src/evm_transition_tool/transition_tool.py [hunk 1]
```diff
@@ -407,3 +437,64 @@ def calc_state_root(
         if state_root is None or not isinstance(state_root, str):
             raise Exception("Unable to calculate state root")
         return new_alloc, bytes.fromhex(state_root[2:])
+
+    def test_fixture(
```
[the remark this answers is not in the document]
Comment (danceratopz) [reply src/evm_transition_tool/transition_tool.py:441]: `verify_fixture` is much better.

File: src/pytest_plugins/test_filler/test_filler.py [hunk 2]
```diff
@@ -196,98 +257,147 @@ def base_test_config(request) -> BaseTestConfig:
...
+        Runs `evm [state|block]test` on each fixture.
+        """
+        for fixture_path, fixture_format in self.json_path_to_fixture_type.items():
+            test_dump_dir = self._get_test_dump_dir(evm_dump_dir, fixture_path)
+            evm_test.test_fixture(fixture_format, fixture_path, test_dump_dir)
+
+    def _get_test_dump_dir(self, evm_dump_dir: Path, fixture_path: Path) -> Optional[Path]:
+        if evm_dump_dir:
+            relative_dump_dir = fixture_path.relative_to(self.output_dir).with_suffix("")
+            return evm_dump_dir / relative_dump_dir
```
[the remark this answers is not in the document]
Comment (danceratopz) [reply src/pytest_plugins/test_filler/test_filler.py:395]: `pathlib` is really amazing; it'd be nice to refactor the rest of our path handling to use it.

File: src/evm_transition_tool/transition_tool.py [hunk 3]
```diff
@@ -407,3 +437,64 @@ def calc_state_root(
...
+        If a client's `t8n` tool varies from the default behavior, this method
+        should be overridden.
+        """
+        command: list[str] = [str(self.binary)]
+
+        if debug_output_path:
+            command += ["--debug", "--json", "--verbosity", "100"]
+
+        if FixtureFormats.is_state_test(fixture_format):
+            if self.statetest_subcommand:
```
[the remark this answers is not in the document]
Comment (danceratopz) [reply src/evm_transition_tool/transition_tool.py:459]: The intention was that `verify_fixture()` (previously `test_fixture()`) should never get called by `test_filler` pytest plugin using a tool that doesn't support statetest|blocktest. And that if a user does try to configure a tool that doesn't support it, that it fails loudly so it's obvious that fixtures aren't actually being verified.

[the remark this answers is not in the document]
Comment (danceratopz) [reply src/evm_transition_tool/transition_tool.py:459]: I did modify the error handling here though to make the error easier to understand 3ae7d097.

File: src/pytest_plugins/test_filler/test_filler.py [hunk 4]
```diff
@@ -52,6 +53,24 @@ def pytest_addoption(parser):
...
+        dest="evm_test_bin",
+        type=Path,
+        default=None,
+        help=(
+            "Path to an evm executable that provides the `blocktest` command. "
+            "Default: The first (geth) 'evm' entry in PATH."
+        ),
+    )
+    evm_group.addoption(
+        "--enable-evm-test",
```
Comment (danceratopz) [inline src/pytest_plugins/test_filler/test_filler.py:68]:
```suggestion
        "--verify-fixtures",
```

File: src/pytest_plugins/test_filler/test_filler.py [hunk 5]
```diff
@@ -52,6 +53,24 @@ def pytest_addoption(parser):
         default=None,
         help="Collect traces of the execution information from the transition tool.",
     )
+    evm_group.addoption(
+        "--evm-test-bin",
```
Comment (danceratopz) [inline src/pytest_plugins/test_filler/test_filler.py:57]:
```suggestion
        "--verify-fixtures-bin",
```

File: src/evm_transition_tool/transition_tool.py [hunk 6]
```diff
@@ -407,3 +437,68 @@ def calc_state_root(
         if state_root is None or not isinstance(state_root, str):
             raise Exception("Unable to calculate state root")
         return new_alloc, bytes.fromhex(state_root[2:])
+
+    def verify_fixture(
```
Comment (danceratopz) [inline src/evm_transition_tool/transition_tool.py:441]: Let's move this to the geth interface and add an abstract method here that raises an Exception if the instantiated class doesn't implement it.

**Review by danceratopz — COMMENTED** [review]
Changes after discussing with Mario.

File: src/pytest_plugins/test_filler/test_filler.py [hunk 7]
```diff
@@ -196,98 +268,185 @@ def base_test_config(request) -> BaseTestConfig:
...
     """
-    Removes the test prefix from a test case name.
+    Converts a test name to a tuple containing the test name and test parameters.
+
+    Example:
+    test_push0_key_sstore[fork=Shanghai] -> test_push0_key_sstore, fork_Shanghai
     """
-    TEST_PREFIX = "test_"
-    if name.startswith(TEST_PREFIX):
-        return name[len(TEST_PREFIX) :]
```
[the remark this answers is not in the document]
Comment (danceratopz) [reply src/pytest_plugins/test_filler/test_filler.py]: Done

**Review by marioevz — APPROVED** [review]
LGTM :)

**Pull request merged**